applications/smart-distancing/smart_distancing/core/_cv_distancing.py

```diff
@@ -1,83 +1,94 @@
-import time
+"""OpenCV Distancing implementation"""
```

Comment:
Aha. The issue is that DeepStream / GStreamer doesn't use OpenCV at all, and it might not even be installed. This is why I added a base class for Distancing with a common interface and some logic in the `core` module level __init__ to decide which implementation gets to become the DefaultDistancing class.

```python
# import Distancing implementations
try:
    # OpenCV might not be installed
    from smart_distancing.core._cv_distancing import *
except ImportError:
    CvDistancing = None
try:
    from smart_distancing.core._ds_distancing import *
except ImportError:  # (DeepStream not found)
    DsDistancing = None
if CvDistancing is None and DsDistancing is None:
    raise ImportError("No available engines (OpenCV or DeepStream)")
# prefer DsDistancing if both are available (NVIDIA platforms)
if DsDistancing:
    DefaultDistancing = DsDistancing
else:
    DefaultDistancing = CvDistancing
```
The Implementations will probably have to differ quite a bit becuase of how they work internally, so a lot of the code can't be re-used. Likely the UI template will have to be separate as well to use a video tag.

The actual overlay drawing likewise has to be reimplemented, since in DeepStream, `nvosd` handles that. The logic for distance calculation can probably be reused (BaseDistancing), and I plan on doing so, but not the actual bbox drawing code.

If you're curious, the actual callback to draw the bounding boxes will likely look something like this:
```
def osd_sink_pad_buffer_probe(pad,info,u_data):
    frame_number=0
    #Intiallizing object counter with 0.
    obj_counter = {
        PGIE_CLASS_ID_VEHICLE:0,
        PGIE_CLASS_ID_PERSON:0,
        PGIE_CLASS_ID_BICYCLE:0,
        PGIE_CLASS_ID_ROADSIGN:0
    }
    num_rects=0

    gst_buffer = info.get_buffer()
    if not gst_buffer:
        print("Unable to get GstBuffer ")
        return

    # Retrieve batch metadata from the gst_buffer
    # Note that pyds.gst_buffer_get_nvds_batch_meta() expects the
    # C address of gst_buffer as input, which is obtained with hash(gst_buffer)
    batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(gst_buffer))
    l_frame = batch_meta.frame_meta_list
    while l_frame is not None:
        try:
            # Note that l_frame.data needs a cast to pyds.NvDsFrameMeta
            # The casting is done by pyds.glist_get_nvds_frame_meta()
            # The casting also keeps ownership of the underlying memory
            # in the C code, so the Python garbage collector will leave
            # it alone.
            #frame_meta = pyds.glist_get_nvds_frame_meta(l_frame.data)
            frame_meta = pyds.NvDsFrameMeta.cast(l_frame.data)
        except StopIteration:
            break

        frame_number=frame_meta.frame_num
        num_rects = frame_meta.num_obj_meta
        l_obj=frame_meta.obj_meta_list
        while l_obj is not None:
            try:
                # Casting l_obj.data to pyds.NvDsObjectMeta
                #obj_meta=pyds.glist_get_nvds_object_meta(l_obj.data)
                obj_meta=pyds.NvDsObjectMeta.cast(l_obj.data)
            except StopIteration:
                break
            obj_counter[obj_meta.class_id] += 1
            obj_meta.rect_params.border_color.set(0.0, 0.0, 1.0, 0.0)
            try: 
                l_obj=l_obj.next
            except StopIteration:
                break

        # Acquiring a display meta object. The memory ownership remains in
        # the C code so downstream plugins can still access it. Otherwise
        # the garbage collector will claim it when this probe function exits.
        display_meta=pyds.nvds_acquire_display_meta_from_pool(batch_meta)
        display_meta.num_labels = 1
        py_nvosd_text_params = display_meta.text_params[0]
        # Setting display text to be shown on screen
        # Note that the pyds module allocates a buffer for the string, and the
        # memory will not be claimed by the garbage collector.
        # Reading the display_text field here will return the C address of the
        # allocated string. Use pyds.get_string() to get the string content.
        py_nvosd_text_params.display_text = "Frame Number={} Number of Objects={} Vehicle_count={} Person_count={}".format(frame_number, num_rects, obj_counter[PGIE_CLASS_ID_VEHICLE], obj_counter[PGIE_CLASS_ID_PERSON])

        # Now set the offsets where the string should appear
        py_nvosd_text_params.x_offset = 10
        py_nvosd_text_params.y_offset = 12

        # Font , font-color and font-size
        py_nvosd_text_params.font_params.font_name = "Serif"
        py_nvosd_text_params.font_params.font_size = 10
        # set(red, green, blue, alpha); set to White
        py_nvosd_text_params.font_params.font_color.set(1.0, 1.0, 1.0, 1.0)

        # Text background color
        py_nvosd_text_params.set_bg_clr = 1
        # set(red, green, blue, alpha); set to Black
        py_nvosd_text_params.text_bg_clr.set(0.0, 0.0, 0.0, 1.0)
        # Using pyds.get_string() to get display_text as string
        print(pyds.get_string(py_nvosd_text_params.display_text))
        pyds.nvds_add_display_meta_to_frame(frame_meta, display_meta)
        try:
            l_frame=l_frame.next
        except StopIteration:
            break
			
    return Gst.PadProbeReturn.OK	
```
This is from nvidia's test [examples](https://github.com/NVIDIA-AI-IOT/deepstream_python_apps/blob/master/apps/deepstream-test1/deepstream_test_1.py#L41).

If you're thinking "OMG, that is ugly AF" I completely agree. GStreamer is an ugly, antiquated thing written in GObject C (no amount of _Python_* makes it pretty), but what it lacks in prettiness it makes up with in speed. Fully optimized, I expect a [single Nano to be able to monitor ~4-8 cameras](https://www.youtube.com/watch?v=Y43W04sMK7I) (probably more if you don't care about dropping frames).

*but there are languages [that do](https://valadoc.org/gstreamer-1.0/Gst.Element.html).

Reply:
the only OMG I have is that OMG I have no idea how DeepStream works :))
I will learn more and if this is what has to be done, let's do it and change it later if we could design a shared interface independent to all of this. 
And yeah, Gstreamer is ugly :D

Reply:
Yeah. I've actually found learning GStreamer to be more complicated than all the other languages I've ever learned combined. It's a massive mess of callback based GObject C. C was not designed with object oriented programming in mind (or async) so...

something sane like...
```python
foo.bar = 123
```
... becomes ...
```C
prefix_foo_set_property('bar', 123, NULL);
// must be null terminated or bad things happen
```
... in GObject C and in python the bindings look like
```python
foo.set_property('bar', 123)
```
or
```python
foo.props.bar = 123
# the reason for 'props' is to avoid namespace conflicts
```
I mentioned earlier there are languages that make this pretty by writing the C for you (Vala, Genie), but they're rarely used outside the GLib/Gnome ecosystem.

Reply:
@mhejerati

Well. The caveat is it can do that with a simple model while using a tracker to only do full inferences on every few frames. Anyway about to head to bed. Gnight and I will continue looking at your comments tomorrow!